src/Cooked/MockChain/BlockChain.hs

```diff
@@ -110,9 +111,27 @@ data MockChainError where
   FailWith :: String -> MockChainError
   deriving (Show, Eq)
 
+-- * MockChain logs
+
+-- | This represents the specific events that should be logged when processing
+-- transactions. If a new kind of event arises, then a new constructor should be
+-- provided here.
+data MockChainLogEntry where
+  -- | Logging a Skeleton as it is submitted by the user.
+  MCLogSubmittedTxSkel :: SkelContext -> TxSkel -> MockChainLogEntry
+  -- | Logging a Skeleton as it has been adjusted by the balancing mechanism,
+  -- alongside fee, collateral utxos and return collateral wallet.
+  MCLogAdjustedTxSkel :: SkelContext -> TxSkel -> Integer -> Set Api.TxOutRef -> Wallet -> MockChainLogEntry
+  -- | Logging the appearance of a new transaction, after a skeleton has been
+  -- successfully sent for validation.
+  MCLogNewTx :: Api.TxId -> MockChainLogEntry
+  -- | Logging the fact that utxos provided by the user for balancing have to be
+  -- discarded for a specific reason.
+  MCLogDiscardedUtxos :: Integer -> String -> MockChainLogEntry
+
```

Comment:
* there used to have one but I removed it, since our error will also stop the mockchain run and thus appear in the printing of the end result. I thought these were duplicates. We can reintroduce them later on when we allow transaction to fail, to do stuff with collaterals.
* GADT: it's not needed indeed. Should I change it? I don't think it bears any harm.

Reply:
1. Even if, for now, it is the last entry of the log (because it stops the mockchain), I still think this is relevant to have. Would it break something? I find it very weird that among all the events that we can log about, mockchain errors are not one of them. It would not have surprised me if this PR changed the return type of running the mockchain to `(Log, Maybe (result, MockChainState))`. I think that it deserves to be an entry already, and if you prefer not to, I think there should be a comment explaining that this is only temporary until we allow to pursue running a trace in which a transaction failed.

2. I think you should change it. Once again, I think we should stick to vanilla Haskell when there is no need to do otherwise. In the future, if we want to study our usage of language extensions in order to clean things up, we will want to know where it is really used, and not have false positives. If leaving it doesn't bear much harm (GADTs are almost standard in practice), I think changing it to vanilla bear even less harm.

Reply:
* Actually, I was mistaken, no such mechanism existed. We were only logging failures for some reason, which is a single instance of errors. I will leave it as it is, because there is a clear distinction between what we log at the moment and what causes an error. We can rethink this distinction later on.
* I have changed it both for mockchain errors and mockchain logs

Reply:
We could even merge errors and logs with logs associated to levels of severity.